fuse/file.go

```diff
@@ -3,103 +3,295 @@ package fuse
+			// for entries with unknown content size.
+			return err
 		}
-		fh.r = updatedEntry
+		resp.Data = data
+	} else {
+		fuseutil.HandleRead(req, resp, f.data)
 	}
 
-	if plugin.WriteAction().IsSupportedOn(updatedEntry) {
-		fh.w = updatedEntry.(plugin.Writable)
+	activity.Record(ctx, "FUSE: Read %v/%v bytes starting at %v from %v", len(resp.Data), req.Size, req.Offset, f)
+	return nil
+}
+
+var _ = fs.HandleWriter(&file{})
+
+func (f *file) Write(ctx context.Context, req *fuse.WriteRequest, resp *fuse.WriteResponse) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+
+	// Ensure handle is in list of writers.
+	f.writers[req.Handle] = struct{}{}
+
+	// If starting write beyond the current length, read to fill it in.
+	if start := int64(len(f.data)); req.Offset > start {
+		data, err := f.load(ctx, start, req.Offset)
+		if err != nil {
+			return err
+		}
+		f.data = append(f.data, data...)
 	}
 
-	if fh.r != nil || fh.w != nil {
-		return &fh, nil
+	// Expand the buffer if necessary and update known size.
+	newLen := req.Offset + int64(len(req.Data))
+	if newLen := int(newLen); newLen > len(f.data) {
+		f.data = append(f.data, make([]byte, newLen-len(f.data))...)
+	}
+	// Write-only entries are assumed not to have a size.
+	if plugin.ReadAction().IsSupportedOn(f.entry) && f.size < uint64(newLen) {
+		f.size = uint64(newLen)
 	}
 
-	activity.Record(ctx, "FUSE: Open unsupported on %v", f)
-	return nil, fuse.ENOTSUP
+	resp.Size = copy(f.data[req.Offset:], req.Data)
+	activity.Record(ctx, "FUSE: Write %v/%v bytes starting at %v from %v", resp.Size, len(req.Data), req.Offset, f)
+	return nil
 }
 
-type fileHandle struct {
-	// This is expected to implement either plugin.Readable or plugin.BlockReadable
-	r  plugin.Entry
-	w  plugin.Writable
-	id string
+func (f *file) load(ctx context.Context, start, end int64) ([]byte, error) {
+	if !plugin.ReadAction().IsSupportedOn(f.entry) {
+		activity.Record(ctx, "FUSE: Non-contiguous writes (at %v) unsupported on %v", start, f)
+		return nil, fuse.ENOTSUP
+	}
+
+	return plugin.Read(ctx, f.entry, end-start, start)
 }
 
-var _ fs.Handle = (*fileHandle)(nil)
-var _ = fs.HandleReleaser(fileHandle{})
-var _ = fs.HandleReader(fileHandle{})
-var _ = fs.HandleWriter(fileHandle{})
+var _ = fs.HandleFlusher(&file{})
+
+// Note that this implementation of Flush only calls plugin.Write if there were previous calls to
+// Write or Setattr. It doesn't check whether the data that's there matches what we're writing.
+func (f *file) Flush(ctx context.Context, req *fuse.FlushRequest) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Flush %v: %+v", f, *req)
+	var releasedWriter bool
 
-func (fh fileHandle) Release(ctx context.Context, req *fuse.ReleaseRequest) error {
-	activity.Record(ctx, "FUSE: Release %v", fh.id)
-	if closer, ok := fh.r.(io.Closer); ok {
-		return closer.Close()
+	// If this handle had an open writer, write current data.
+	if _, ok := f.writers[req.Handle]; ok {
```

Comment:
It's not uncommon to call Flush on a handle opened for reading (not sure why). It would also not have an open writer if you opened it for writing but didn't do any calls `write` calls.